sandman/source/shell_attr.h

```diff
@@ -0,0 +1,392 @@
+		{
+			static_assert(sizeof...(args) > 0u,
+							  "Calling this function without any arguments is probably a mistake.");
+
+			return Wrapper(this->BuildAttr(), std::forward<ParamsT>(args)...);
+		}
+	};
+
+	[[nodiscard]] constexpr
+	Attr::BackgroundColor operator|(Attr const attributes, Attr::BackgroundColor const color)
+	{
+		return color | attributes;
+	}
+
+	struct Attr::ColorPair
+	{
+		Attr m_Ancillary;
+
+		// Bit manipulations rely on the assumption that this assertion is true.
+		static_assert([]() constexpr -> bool {
+			ColorMatrix::Index maxColorIndex{ColorMatrix::kList.at(0u)};
+
+			for (ColorMatrix::Index const colorIndex : ColorMatrix::kList)
+			{
+				maxColorIndex = std::max(colorIndex, maxColorIndex);
+			}
+
+			return ::Common::Enum::IntCast(maxColorIndex) == 0b111u;
+		}());
+
+		// Two color pair indices are packed into this value.
+		std::underlying_type_t<ColorMatrix::Index> m_ColorIndexPair;
+
+		[[nodiscard]] constexpr ColorPair operator|(Attr const attributes) const
+		{
+			return { this->m_Ancillary | attributes, m_ColorIndexPair };
+		}
+
+		[[nodiscard]] constexpr Attr BuildAttr() const
+		{
+			using namespace ColorMatrix;
+
+			ForegroundIndex const foregroundColor(
+				static_cast<Index>(m_ColorIndexPair bitand 0b000111u));
+
+			BackgroundIndex const backgroundColor(
+				static_cast<Index>((m_ColorIndexPair bitand 0b111000u) >> 3u));
+
+			Attr const colorPair{ GetPair(foregroundColor, backgroundColor) };
+
+			return m_Ancillary | colorPair;
+		}
+
+		template <typename... ParamsT>
+		[[nodiscard]] constexpr Wrapper<ParamsT&&...> operator()(ParamsT&&... args) const
+		{
+			static_assert(sizeof...(args) > 0u,
+							  "Calling this function without any arguments is probably a mistake.");
+
+			return Wrapper(this->BuildAttr(), std::forward<ParamsT>(args)...);
+		}
+	};
+
+	[[nodiscard]] constexpr
+	Attr::ColorPair operator|(Attr const attributes, Attr::ColorPair const colorPair)
+	{
+		return colorPair | attributes;
+	}
+
+	[[nodiscard]] constexpr
+	Attr::ColorPair operator|(Attr::ForegroundColor const foregroundColor,
+									  Attr::BackgroundColor const backgroundColor)
+	{
+		using Common::Enum::IntCast;
+
+		std::underlying_type_t<ColorMatrix::Index> const colorIndexPair
+		{
+			static_cast<std::underlying_type_t<ColorMatrix::Index>>
+			(
+				IntCast(foregroundColor.m_ColorIndex) bitor
+				(IntCast(backgroundColor.m_ColorIndex) << 3u)
+			)
+		};
+
+		return { foregroundColor.m_Ancillary | backgroundColor.m_Ancillary, colorIndexPair };
+	}
+
+	[[nodiscard]] constexpr
+	Attr::ColorPair operator|(Attr::BackgroundColor const backgroundColor,
+									  Attr::ForegroundColor const foregroundColor)
+	{
+		return foregroundColor | backgroundColor;
+	}
+
+	inline namespace CharacterAttributeObjectWrapperFactories
+	{
+
+		// NOLINTBEGIN(readability-identifier-naming)
+
+		// Character attribute.
+		inline constexpr Attr
```

Comment:
Should I also break these long lines? They go over the 100-column limit, but I didn't want to break them because the symmetry between all the declarations looked aesthetically pleasing.

Reply:
Resolved in 973e9785971671e947305a6949496cc912f813d2.